File: app/src/main/java/com/boardgamegeek/ui/dialog/SliderFilterDialog.java
package com.boardgamegeek.ui.dialog;

import android.content.Context;
import android.content.DialogInterface;
import android.support.annotation.StringRes;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.CheckBox;
import android.widget.TextView;

import com.appyvet.rangebar.RangeBar;
import com.appyvet.rangebar.RangeBar.OnRangeBarChangeListener;
import com.appyvet.rangebar.RangeBar.PinTextFormatter;
import com.boardgamegeek.R;
import com.boardgamegeek.filterer.CollectionFilterer;
import com.boardgamegeek.interfaces.CollectionView;
import com.boardgamegeek.util.MathUtils;
import com.boardgamegeek.util.StringUtils;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public abstract class SliderFilterDialog implements CollectionFilterDialog {
	private Integer low;
	private Integer high;
	@BindView(R.id.explanation) TextView explanationView;
	@BindView(R.id.checkbox) CheckBox checkBox;
	@BindView(R.id.range_bar) RangeBar rangeBar;

	public void createDialog(final Context context, final CollectionView view, CollectionFilterer filter) {
		LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
		View layout = inflater.inflate(R.layout.dialog_slider_filter, null);
		ButterKnife.bind(this, layout);

		InitialValues initialValues = initValues(filter);
		low = MathUtils.constrain(initialValues.min, getAbsoluteMin(), getAbsoluteMax());
		high = MathUtils.constrain(initialValues.max, getAbsoluteMin(), getAbsoluteMax());

		initSlider();

		checkBox.setVisibility(getCheckboxVisibility());
		checkBox.setText(getCheckboxTextId());
		checkBox.setChecked(initialValues.isChecked);

		initExplanation();

		AlertDialog.Builder builder = new AlertDialog.Builder(context).setTitle(getTitleId())
			.setNegativeButton(R.string.clear, new DialogInterface.OnClickListener() {
				@Override
				public void onClick(DialogInterface dialog, int which) {
					view.removeFilter(getType(context));
				}
			}).setPositiveButton(R.string.set, new DialogInterface.OnClickListener() {
				@Override
				public void onClick(DialogInterface dialog, int id) {
					view.addFilter(getPositiveData(context, low, high, checkBox.isChecked()));
				}
			}).setView(layout);

		builder.create().show();
	}

	private void initSlider() {
		rangeBar.setTickStart(getAbsoluteMin());
		rangeBar.setTickEnd(getAbsoluteMax());
		rangeBar.setRangePinsByValue(low, high);

		rangeBar.setPinTextFormatter(new PinTextFormatter() {
			@Override
			public String getText(String value) {
				return getPinText(value);
			}
		});

		rangeBar.setOnRangeBarChangeListener(new OnRangeBarChangeListener() {
			@Override
			public void onRangeChangeListener(RangeBar rangeBar, int leftPinIndex, int rightPinIndex, String leftPinValue, String rightPinValue) {
				low = getPinValue(leftPinValue);
				high = getPinValue(rightPinValue);
			}
		});
	}

	@OnClick(R.id.min_up)
	public void onMinUpClick() {
		if (rangeBar.getLeftIndex() < rangeBar.getTickCount() - 1) {
			if (rangeBar.getLeftIndex() == rangeBar.getRightIndex()) {
				updateRange(rangeBar.getLeftIndex() + 1, rangeBar.getRightIndex() + 1);
			} else {
				updateRange(rangeBar.getLeftIndex() + 1, rangeBar.getRightIndex());
			}
		}
	}

	@OnClick(R.id.min_down)
	public void onMinDownClick() {
		if (rangeBar.getLeftIndex() > 0) {
			updateRange(rangeBar.getLeftIndex() - 1, rangeBar.getRightIndex());
		}
	}

	@OnClick(R.id.max_up)
	public void onMaxUpClick() {
		if (rangeBar.getRightIndex() < rangeBar.getTickCount() - 1) {
			updateRange(rangeBar.getLeftIndex(), rangeBar.getRightIndex() + 1);
		}
	}

	@OnClick(R.id.max_down)
	public void onMaxDownClick() {
		if (rangeBar.getRightIndex() > 0) {
			if (rangeBar.getLeftIndex() == rangeBar.getRightIndex()) {
				updateRange(rangeBar.getLeftIndex() - 1, rangeBar.getRightIndex() - 1);
			} else {
				updateRange(rangeBar.getLeftIndex(), rangeBar.getRightIndex() - 1);
			}
		}
	}

	private void updateRange(int leftPinIndex, int rightIndex) {
		rangeBar.setRangePinsByIndices(leftPinIndex, rightIndex);
		// HACK to make the pins remain visible
		rangeBar.setLeft(rangeBar.getLeft() + 1);
	}

	private void initExplanation() {
		if (getDescriptionId() == -1) {
			explanationView.setVisibility(View.GONE);
		} else {
			explanationView.setText(getDescriptionId());
			explanationView.setVisibility(View.VISIBLE);
		}
	}

	protected abstract InitialValues initValues(CollectionFilterer filter);

	@StringRes
	protected abstract int getTitleId();

	protected abstract CollectionFilterer getPositiveData(final Context context, int min, int max, boolean checkbox);

	protected int getCheckboxVisibility() {
		return View.VISIBLE;
	}

	@StringRes
	protected int getCheckboxTextId() {
		return R.string.include_missing_values;
	}

	@StringRes
	protected int getDescriptionId() {
		return -1;
	}

	protected abstract int getAbsoluteMin();

	protected abstract int getAbsoluteMax();

	protected String getPinText(String value) {
		return value;
	}

	protected int getPinValue(String text) {
		return StringUtils.parseInt(text);
	}

	class InitialValues {
		final int min;
		final int max;
		final boolean isChecked;

		InitialValues(int min, int max) {
			this.min = min;
			this.max = max;
			this.isChecked = false;
		}

		InitialValues(int min, int max, boolean isChecked) {
			this.min = min;
			this.max = max;
			this.isChecked = isChecked;
		}
	}
}

File: app/src/main/java/com/boardgamegeek/ui/dialog/SliderFilterDialog.kt
package com.boardgamegeek.ui.dialog

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import com.appyvet.materialrangebar.RangeBar
import com.boardgamegeek.R
import com.boardgamegeek.databinding.DialogSliderFilterBinding
import com.boardgamegeek.filterer.CollectionFilterer

abstract class SliderFilterDialog : CollectionFilterDialog {
    private var low: Int = 0
    private var high: Int = 0

    @get:StringRes
    protected abstract val titleResId: Int

    protected open val checkboxVisibility: Int = View.VISIBLE

    protected open val checkboxTextResId: Int
        @StringRes
        get() = R.string.include_missing_values

    protected open val supportsSlider = true

    protected open val rangeInterval = 1

    protected open val descriptionResId: Int
        @StringRes
        get() = INVALID_STRING_RES_ID

    protected abstract val absoluteMin: Int

    protected abstract val absoluteMax: Int

    override fun createDialog(context: Context, listener: CollectionFilterDialog.OnFilterChangedListener?, filter: CollectionFilterer?) {
        val binding = DialogSliderFilterBinding.inflate(LayoutInflater.from(context))

        val initialValues = initValues(filter)
        low = initialValues.min.coerceIn(absoluteMin, absoluteMax)
        high = initialValues.max.coerceIn(absoluteMin, absoluteMax)

        binding.rangeBar.apply {
            tickStart = absoluteMin.toFloat()
            tickEnd = absoluteMax.toFloat()
            setRangePinsByValue(low.toFloat(), high.toFloat())
            if (low == high && supportsSlider) setRangeBarEnabled(false)
            setPinTextFormatter { value -> getPinText(context, value) }
            setOnRangeBarChangeListener(object : RangeBar.OnRangeBarChangeListener {
                override fun onRangeChangeListener(rangeBar: RangeBar?, leftPinIndex: Int, rightPinIndex: Int, leftPinValue: String?, rightPinValue: String?) {
                    high = (rightPinIndex + absoluteMin).coerceIn(absoluteMin, absoluteMax)
                    low = if (isRangeBar) {
                        (leftPinIndex + absoluteMin).coerceIn(absoluteMin, absoluteMax)
                    } else {
                        high
                    }
                }

                override fun onTouchStarted(rangeBar: RangeBar?) {
                }

                override fun onTouchEnded(rangeBar: RangeBar?) {
                }
            })
        }

        binding.minDownButton.setOnClickListener {
            binding.rangeBar.apply {
                if (leftIndex > 0) {
                    updateRange(this, leftIndex - 1, rightIndex)
                }
            }
        }

        binding.minUpButton.setOnClickListener {
            binding.rangeBar.apply {
                if (leftIndex < tickCount - 1) {
                    if (leftIndex == rightIndex) {
                        updateRange(this, leftIndex + 1, rightIndex + 1)
                    } else {
                        updateRange(this, leftIndex + 1, rightIndex)
                    }
                }
            }
        }

        binding.maxUpButton.setOnClickListener {
            binding.rangeBar.apply {
                if (rightIndex < tickCount - 1) {
                    updateRange(this, leftIndex, rightIndex + 1)
                }
            }
        }

        binding.maxDownButton.setOnClickListener {
            binding.rangeBar.apply {
                if (rightIndex > 0) {
                    if (leftIndex == rightIndex) {
                        updateRange(this, leftIndex - 1, rightIndex - 1)
                    } else {
                        updateRange(this, leftIndex, rightIndex - 1)
                    }
                }
            }
        }

        binding.rangeCheckBox.apply {
            visibility = if (supportsSlider) View.VISIBLE else View.GONE
            isChecked = (low != high)
            setOnCheckedChangeListener { _, isChecked ->
                binding.rangeBar.setRangeBarEnabled(isChecked)
                binding.minDownButton.visibility = if (isChecked) View.VISIBLE else View.GONE
                binding.minUpButton.visibility = if (isChecked) View.VISIBLE else View.GONE
                binding.buttonSpace.visibility = if (isChecked) View.VISIBLE else View.GONE
                if (isChecked) {
                    binding.rangeBar.apply {
                        if (leftIndex == rightIndex) {
                            if (leftIndex > 0) {
                                updateRange(this, leftIndex - rangeInterval, rightIndex)
                            } else {
                                updateRange(this, leftIndex, rightIndex + rangeInterval)
                            }
                        } else {
                            updateRange(this, leftIndex, rightIndex)
                        }
                    }
                } else {
                    binding.rangeBar.apply { updateRange(this, leftIndex, rightIndex) }
                }
            }
        }

        binding.checkBox.apply {
            visibility = checkboxVisibility
            setText(checkboxTextResId)
            isChecked = initialValues.isChecked
        }

        binding.explanationView.apply {
            visibility = if (descriptionResId == INVALID_STRING_RES_ID) {
                View.GONE
            } else {
                setText(descriptionResId)
                View.VISIBLE
            }
        }

        val builder = AlertDialog.Builder(context, R.style.Theme_bgglight_Dialog_Alert)
                .setTitle(titleResId)
                .setNegativeButton(R.string.clear) { _, _ -> listener?.removeFilter(getType(context)) }
                .setPositiveButton(R.string.set) { _, _ -> listener?.addFilter(getPositiveData(context, low, high, binding.checkBox.isChecked)) }
                .setView(binding.root)

        builder.create().show()
    }

    private fun updateRange(rangeBar: RangeBar, leftPinIndex: Int, rightPinIndex: Int) {
        rangeBar.setRangePinsByIndices(
                leftPinIndex.coerceIn(0, rangeBar.tickCount - 1),
                rightPinIndex.coerceIn(0, rangeBar.tickCount - 1)
        )
        // HACK to make the pins remain visible when using the up/down buttons
        rangeBar.left = rangeBar.left + 1
    }

    protected abstract fun initValues(filter: CollectionFilterer?): InitialValues

    protected abstract fun getPositiveData(context: Context, min: Int, max: Int, checkbox: Boolean): CollectionFilterer

    protected open fun getPinText(context: Context, value: String) = value

    data class InitialValues @JvmOverloads constructor(val min: Int, val max: Int, val isChecked: Boolean = false)

    companion object {
        private const val INVALID_STRING_RES_ID = -1
    }
}
